File: livingrimoire start here/LivinGrimoire java/LivinGrimoire 030123/LivinGrimoire core/Algorithm.java
package LivinGrimoire;

import java.util.ArrayList;

// a step-by-step plan to achieve a goal
public class Algorithm {
    private String goal;
    private String representation;
    private ArrayList<Mutatable> algParts = new ArrayList<>();

    public Algorithm(String goal, String representation, ArrayList<Mutatable> algParts) {
        super();
        this.goal = goal.isEmpty()? "unknown": goal;
        this.representation = representation.isEmpty()? "unknown":representation;
        this.algParts = algParts;
    }
    // *constract with string and goal

    public String getGoal() {
        return goal;
    }

    public String getRepresentation() {
        return representation;
    }

    public ArrayList<Mutatable> getAlgParts() {
        return algParts;
    }

    public int getSize() {
        return algParts.size();
    }

    public Algorithm clone() {
        // returns a deep copy algorithm
        ArrayList<Mutatable> parts = new ArrayList<>();
        for (Mutatable absAlgPart : this.algParts) {
            parts.add(absAlgPart.clone());
        }
        return new Algorithm(this.goal, getRepresentation(), parts);
    }
}

File: livingrimoire start here/LivingGrimoire Kotlin/lgkt/Algorithm.kt
package lgkt


// a step-by-step plan to achieve a goal
class Algorithm(goal: String, representation: String, algParts: ArrayList<Mutatable>) {
    // *constract with string and goal
    val goal: String
    val representation: String
    var algParts = ArrayList<Mutatable>()

    init {
        this.goal = if (goal.isEmpty()) "unknown" else goal
        this.representation = if (representation.isEmpty()) "unknown" else representation
        this.algParts = algParts
    }

    val size: Int
        get() = algParts.size

    fun clone(): Algorithm {
        // returns a deep copy algorithm
        val parts = ArrayList<Mutatable>()
        for (absAlgPart in algParts) {
            parts.add(absAlgPart.clone())
        }
        return Algorithm(goal, representation, parts)
    }
}